src/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/RandomNumberGeneratorImplementation.cs

```diff
@@ -0,0 +1,58 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+namespace System.Security.Cryptography
+{
+    internal sealed partial class RandomNumberGeneratorImplementation : RandomNumberGenerator
+    {
+        public override void GetNonZeroBytes(byte[] data)
+        {
+            if (data == null)
+                throw new ArgumentNullException(nameof(data));
+
+            GetBytes(data);
+
+            int indexOfFirst0Byte = data.Length;
+            for (int i = 0; i < data.Length; i++)
+            {
+                if (data[i] == 0)
+                {
+                    indexOfFirst0Byte = i;
+                    break;
+                }
+            }
+
+            for (int i = indexOfFirst0Byte; i < data.Length; i++)
+            {
+                if (data[i] != 0)
+                {
+                    data[indexOfFirst0Byte++] = data[i];
+                }
+            }
+
+            while (indexOfFirst0Byte < data.Length)
+            {
+                // This should be more than enough to fill the rest in one iteration
+                byte[] tmp = new byte[2 * (data.Length - indexOfFirst0Byte)];
```

Comment:
The original code may be trying to minimize the number of writes to the output (i.e. because it fills in 'data' first with the bytes instead of filling a temp buffer first and then copying to data) and also perhaps minimizing the number of allocations + calls to get random data if there's some padding to fill (i.e. 2*bytesremaining) because it is more efficient to do it once than do it 2+ more times in smaller chunks. For example, with a large buffer of 65536 it would on a 'perfect' distribution have to allocate\get random numbers 3 times if 2* was not used (65536[]->256[] -> 1[]) instead of 2 times (65536[]->512[]).